### PR TITLE
Better handling of PoolClearedError during legacy read

### PR DESCRIPTION
**Pull request by felixmo**
Based on https://github.com/braze-inc/mongo-ruby-driver/blob/deployed-in-platform/spec/integration/retryable_reads_errors_spec.rb#L76 it looks like `PoolClearedError` can be retried on reads.

Also updates `#legacy_read_with_retry?` to check whether the error responds to `#retryable?` before attempting to call it.

### CODE REVIEW

File: lib/mongo/retryable/read_worker.rb [hunk 1]
```diff
@@ -218,7 +218,7 @@ def legacy_read_with_retry(session, server_selector, &block)
         
         if is_retryable_exception?(e)
           raise e if attempt > client.max_read_retries || session&.in_transaction?
-        elsif e.retryable? && !session&.in_transaction?
+        elsif e.respond_to?(:retryable?) && e.retryable? && !session&.in_transaction?
```
Comment (joelim41) [inline lib/mongo/retryable/read_worker.rb:221]: should we match the new method [modern_read_with_retry](https://github.com/braze-inc/mongo-ruby-driver/blob/6322113530ca1619e5e4fee05135c5a7e2784231/lib/mongo/retryable/read_worker.rb#L198) way of handling these new errors instead? e.g. handle write_retryable errors here

File: lib/mongo/retryable/base_worker.rb [hunk 2]
```diff
@@ -58,7 +58,8 @@ def retryable_exceptions
           Error::ConnectionPerished,
           Error::ServerNotUsable,
           Error::SocketError,
-          Error::SocketTimeoutError
+          Error::SocketTimeoutError,
+          Error::PoolClearedError,
```
Comment (joelim41) [inline lib/mongo/retryable/base_worker.rb:62]: might want to add the other PoolErrors here

File: lib/mongo/retryable/read_worker.rb (hunk 1 above)
Comment (felixmo) [reply lib/mongo/retryable/read_worker.rb:221]: I tried that initially but saw some legit test failures so I think that might be changing the behavior of this too much

**Comment by jasonpenny** [comment]
I don't feel confident that any of us know this code well enough to be patching this in a rush.

I think we ought to open a PR upstream and to get more perspective from the maintainers

**Pull request closed without merging**